build.gradle

```diff
@@ -148,6 +148,19 @@ allprojects {
             "-Xopt-in=kotlin.RequiresOptIn"
         ]
     }
+
+    // FIXME(bug 1726765): Temporary hack to ensure there's a single version of Glean in use across the app.
+    configurations.all {
```

Comment:
Testing this change with https://github.com/mozilla-mobile/android-components/pull/10828 locally and verified I only have a single version of Glean (40.0.0). Verified using `./gradlew app:dependencies`.

However, I see this check failing with:

```
Could not find method AssertionError() for arguments [Cannot have a conflict on Glean requested] on org.mozilla.telemetry:glean:40.0.0 of type org.gradle.api.internal.artifacts.result.DefaultResolvedComponentResult.

Possible solution:
 - Declare repository providing the artifact, see the documentation at https://docs.gradle.org/current/userguide/declaring_repositories.html
```

@badboy Does running a `Sync Project with Gradle` in Android Studio or just running `./gradlew app:dependencies` with this check commented in work for you having a local A-C version with your changes?

Reply:
After commenting out this validation my build (both command line and Android Studio) runs into a `StackOverflowError`: 

```
[Glean Plugin] Adding Glean capability through GeckoView (provided by geckoview-nightly-omni)

> Task :app:getDependencies FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:getDependencies'.
> java.lang.StackOverflowError (no error message)
```

```
Cause: java.lang.StackOverflowError
	at com.google.common.util.concurrent.SettableFuture.set(SettableFuture.java:48)
	at com.google.common.cache.LocalCache$LoadingValueReference.set(LocalCache.java:3415)
	at com.google.common.cache.LocalCache$LoadingValueReference.notifyNewValue(LocalCache.java:3431)
	at com.google.common.cache.LocalCache$Segment.setValue(LocalCache.java:2015)
	at com.google.common.cache.LocalCache$Segment.storeLoadedValue(LocalCache.java:3074)
	at com.google.common.cache.LocalCache$Segment.getAndRecordStats(LocalCache.java:2234)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2195)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2153)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2043)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3851)
	at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4713)
	at org.gradle.internal.resources.AbstractResourceLockRegistry.getOrRegisterResourceLock(AbstractResourceLockRegistry.java:45)
	at org.gradle.internal.work.DefaultWorkerLeaseService$WorkerLeaseLockRegistry.getResourceLock(DefaultWorkerLeaseService.java:353)
	at org.gradle.internal.work.DefaultWorkerLeaseService.getWorkerLease(DefaultWorkerLeaseService.java:95)
	at org.gradle.internal.work.DefaultWorkerLeaseService.getWorkerLease(DefaultWorkerLeaseService.java:102)
	at org.gradle.internal.work.DefaultWorkerLeaseService.getWorkerLease(DefaultWorkerLeaseService.java:57)
	at org.gradle.internal.operations.DefaultBuildOperationQueue.<init>(DefaultBuildOperationQueue.java:56)
	at org.gradle.internal.operations.DefaultBuildOperationQueueFactory.create(DefaultBuildOperationQueueFactory.java:31)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.executeInParallel(DefaultBuildOperationExecutor.java:119)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$runAll$4(DefaultBuildOperationExecutor.java:100)
	at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.runAll(DefaultBuildOperationExecutor.java:100)
	at org.gradle.api.internal.artifacts.ivyservice.resolveengine.artifact.ParallelResolveArtifactSet$VisitingSet.visit(ParallelResolveArtifactSet.java:65)
	at org.gradle.api.internal.artifacts.DefaultResolvedDependency.sort(DefaultResolvedDependency.java:129)
	at org.gradle.api.internal.artifacts.DefaultResolvedDependency.getModuleArtifacts(DefaultResolvedDependency.java:107)
	at org.gradle.api.internal.artifacts.DefaultResolvedDependency.getAllModuleArtifacts(DefaultResolvedDependency.java:113)
	at org.gradle.api.internal.artifacts.DefaultResolvedDependency.getAllModuleArtifacts(DefaultResolvedDependency.java:115)
	at org.gradle.api.internal.artifacts.DefaultResolvedDependency.getAllModuleArtifacts(DefaultResolvedDependency.java:115)
	at org.gradle.api.internal.artifacts.DefaultResolvedDependency.getAllModuleArtifacts(DefaultResolvedDependency.java:115)
...
```

Reply:
Adding that I tested this with a locally published A-C version based on your [PR](mozilla-mobile/android-components#10828)) to build an actual Nightly release with a released A-C version for testing: https://mozac.org/contributing/testing-components-inside-app (see **manual** flow).

Reply:
I digged into this a little bit, it looks like somehow `glean-native` ends up depending on `geckoview` which creates a cyclic dependency gv -> glean -> glean-native -> gv

from `./gradlew app:dependencies`:

```
ah yeah... look at this:

+--- org.mozilla.geckoview:geckoview-nightly-omni:93.0.20210823092315
|    |    \--- org.mozilla.telemetry:glean:40.0.0
...
|    |         +--- org.mozilla.telemetry:glean-native:40.0.0 -> org.mozilla.geckoview:geckoview-nightly-omni:93.0.20210823092315 (*)
```

Reply:
Or maybe... I might be misunderstanding what the meaning of this arrow is:

```
org.mozilla.telemetry:glean-native:40.0.0 -> org.mozilla.geckoview:geckoview-nightly-omni:93.0.20210823092315 (*)
```

geckoview _provides_ `glean-native` so maybe that's what this is telling us.

Reply:
Yeah nevermind, looking at other dependencies, the `->` seems to mean that the dependency to the left has been substituted with the one on the right, which is correct as we want GV to provide `glean-native`.

Reply:
*BUT* we also have geckoview depending on `glean`, which I think is the problem here? because the deps are now

```
glean -> glean-native = geckoview -> glean
```

Reply:
I think GeckoView should depend on `glean-native` not `glean` and that will solve this. Testing it locally

Reply:
I wonder if somehow the plugin can also remove the `glean -> glean-native` dependency when it detects GV

Reply:
Maybe what we want here is defining capabilities in GeckoView through the Gradle Module Metadata instead of the Glean plugin, that should work and it looks like the only thing required is to upgrade to Gradle 6.0 (in GV): https://docs.gradle.org/current/userguide/component_capabilities.html (see bottom of the page)

Reply:
Thanks for the thorough look. I briefly stumbled across the stackoverflow, but thought it was fixed.
Now looking at agi's comments it's of course clear why it was happening.

Reply:
> So the check is failing. As a side-note, the `AssertionError` is somehow also undefined

D'oh. Yeah, that line needs to be 'throw new AssertionError(...)`. I just didn't notice because ... well, it broke the build anyway.